src/emulator/apphosting/serve.ts

```diff
@@ -167,3 +204,83 @@ export function getEmulatorEnvs(): Record<string, string> {
 
   return envs;
 }
+
+type Dependency = {
+  name: string;
+  version: string;
+  path: string;
+  dependencies?: Record<string, Dependency>;
+};
+
+async function tripFirebasePostinstall(
+  rootDirectory: string,
+  env: NodeJS.ProcessEnv,
+): Promise<void> {
+  const npmLs = spawnSync("npm", ["ls", "@firebase/util", "--json", "--long"], {
+    cwd: rootDirectory,
+    shell: process.platform === "win32",
+  });
+  const npmLsResults = JSON.parse(npmLs.stdout.toString().trim());
+  const dependenciesToSearch: Dependency[] = Object.values(npmLsResults.dependencies);
+  const firebaseUtilPaths: string[] = [];
+  for (const dependency of dependenciesToSearch) {
+    if (
+      dependency.name === "@firebase/util" &&
+      semverGte(dependency.version, "1.11.0") &&
+      firebaseUtilPaths.indexOf(dependency.path) === -1
+    ) {
+      firebaseUtilPaths.push(dependency.path);
+    }
+    if (dependency.dependencies) {
+      dependenciesToSearch.push(...Object.values(dependency.dependencies));
+    }
+  }
+  await Promise.all(
+    firebaseUtilPaths.map(
+      (path) =>
+        new Promise<void>((resolve) => {
+          spawnSync("npm", ["run", "postinstall"], {
+            cwd: path,
+            env,
+            stdio: "ignore",
+            shell: process.platform === "win32",
+          });
+          resolve();
+        }),
+    ),
+  );
+}
+
+async function getBackendAppConfig(
+  projectId?: string,
+  backendId?: string,
+): Promise<WebConfig | undefined> {
+  if (!projectId) {
+    return undefined;
+  }
+
+  if (Constants.isDemoProject(projectId)) {
+    return constructDefaultWebSetup(projectId);
+  }
+
+  if (!backendId) {
+    return undefined;
+  }
+
+  const backendsList = await apphosting.listBackends(projectId, "-").catch(() => undefined);
+  const backend = backendsList?.backends.find((b) => b.name.split("/").pop() === backendId);
```

Comment:
we have a utility for parsing the backend ID from the full resource name called `parseBackendName` in src/apphosting/backend.ts